### PR TITLE
Parquet

### PR DESCRIPTION
**Pull request by bmcdonald3**
This PR adds the ability to read and write int64 Parquet files in Arkouda through a `pdarray.save_parquet()` function and an `ak.read_parquet()` function. These functions behave in much the same way as the `pdarray.save()` and `ak.load()` functions (see docstrings). This is only the initial support, and more dtypes will likely be added as use necessitates.

For reading Parquet files, the Arrow C++ library was chosen due to library completeness and portability. The actual functionality is written in C++, with light C wrappers to allow leveraging of the Chapel C interoperability features, which requires the C++ code to be compiled into an object file prior to calling the functions within Chapel.

On the Chapel side, the functionality is modeled off of the HDF5 code, but currently only supports int64 and int32 Arrow datatypes, which are read into int64 pdarrays.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -136,7 +136,8 @@
         'pyzmq>=20.0.0',
         'typeguard==2.10.0',
         'pyfiglet',
-        'versioneer'
+        'versioneer',
+        'pyarrow'
```
Comment (ronawho) [inline setup.py:140]: What all is `pyarrow` needed for?

Comment (ronawho) [reply setup.py:140]: (If it's only for testing, we should move this to the `extras_require` dependencies list)

Comment (bmcdonald3) [reply setup.py:140]: `pyarrow` is used in the unit tests to compare the Arkouda Parquet functionality to the `pyarrow` parquet functionality. The thought here was to take the same approach as testing against numpy.

File: dep/checkArrow.chpl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require 'src/ArrowInclude.chpl';
+
+use ArrowInclude;
```
Comment (ronawho) [inline dep/checkArrow.chpl:3]: What the double require and use?

File: benchmarks/parquetIO.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    parser.add_argument('hostname', help='Hostname of arkouda server')
+    parser.add_argument('port', type=int, help='Port of arkouda server')
+    parser.add_argument('-n', '--size', type=int, default=10**8, help='Problem size: length of array to write/read')
+    parser.add_argument('-t', '--trials', type=int, default=1, help='Number of times to run the benchmark')
+    parser.add_argument('-d', '--dtype', default='int64', help='Dtype of array ({})'.format(', '.join(TYPES)))
+    parser.add_argument('-p', '--path', default=os.getcwd()+'/ak-pq-test', help='Target path for measuring read/write rates')
+    parser.add_argument('--correctness-only', default=False, action='store_true', help='Only check correctness, not performance.')
+    parser.add_argument('-s', '--seed', default=None, type=int, help='Value to initialize random number generator')
+    return parser
+
```
Comment (ronawho) [inline benchmarks/parquetIO.py:50]: Please add a correctness mode (see IO.py check_correctness)

File: benchmarks/parquetIO.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import time, argparse
+import arkouda as ak
+import os
+from glob import glob
+
+TYPES = ('int64')
```
Comment (ronawho) [inline benchmarks/parquetIO.py:6]: Are any types beyond `int64` supported today? `float64` maybe? If so, we should add those types here

File: arkouda/pdarrayclass.py [hunk 5]
```diff
@@ -1005,6 +1005,82 @@ def save(self, prefix_path : str, dataset : str='array', mode : str='truncate')
...
+
+        Parameters
+        ----------
+        prefix_path : str
+            Directory and filename prefix that all output files share
+        dataset : str
+            Name of the dataset to create in Parquet files (must not already exist)
+        mode : str {'truncate' | 'append'}
+            By default, truncate (overwrite) output files, if they exist.
+            If 'append', attempt to create new dataset in existing files.
```
Comment (ronawho) [inline arkouda/pdarrayclass.py:1024]: Do you support the different modes?

File: arkouda/pdarrayIO.py [hunk 6]
```diff
@@ -111,6 +111,99 @@ def read_hdf(dsetName : str, filenames : Union[str,List[str]],
...
+    strictTypes: bool
+        If True (default), require all dtypes in all files to have the
+        same precision and sign. If False, allow dtypes of different
+        precision and sign across different files. For example, if one 
+        file contains a uint32 dataset and another contains an int64
+        dataset, the contents of both will be read into an int64 pdarray.
+    allow_errors: bool
+        Default False, if True will allow files with read errors to be skipped
+        instead of failing.  A warning will be included in the return containing
+        the total number of files skipped due to failure and up to 10 filenames.
```
Comment (ronawho) [inline arkouda/pdarrayIO.py:136]: Do we do anything with allow_errors on the server side for parquet?

File: arkouda/pdarrayIO.py [hunk 7]
```diff
@@ -111,6 +111,99 @@ def read_hdf(dsetName : str, filenames : Union[str,List[str]],
...
+    filenames : list or str
+        Either a list of filenames or shell expression
+    dsetName : str
+        The name of the dataset (must be the same across all files)
+    strictTypes: bool
+        If True (default), require all dtypes in all files to have the
+        same precision and sign. If False, allow dtypes of different
+        precision and sign across different files. For example, if one 
+        file contains a uint32 dataset and another contains an int64
+        dataset, the contents of both will be read into an int64 pdarray.
```
Comment (ronawho) [inline arkouda/pdarrayIO.py:132]: Just pointing out this comment since I think it points out for other file formats they support reading in different types in the files into a single pdarray. Not sure how this works today or how they figure out the pdarray type then.

File: arkouda/pdarrayIO.py [hunk 8]
```diff
@@ -111,6 +111,99 @@ def read_hdf(dsetName : str, filenames : Union[str,List[str]],
...
+    rep_msg = generic_msg(cmd="readAllParquet", args=
+                          f"{strictTypes} {len(dsetname)} {len(filenames)} {allow_errors} {json.dumps(dsetname)} | {json.dumps(filenames)}")
+    rep = json.loads(rep_msg)  # See GenSymIO._buildReadAllHdfMsgJson for json structure
+    items = rep["items"] if "items" in rep else []
+    file_errors = rep["file_errors"] if "file_errors" in rep else []
+
+    # We have a couple possible return conditions
+    # 1. We have multiple items returned i.e. multi pdarrays, multi strings, multi pdarrays & strings
+    # 2. We have a single pdarray
+    # 3. We have a single strings object
```
Comment (ronawho) [inline arkouda/pdarrayIO.py:185]: Do we support all these options today? (If so, do we have tests for all of them?)

File: benchmarks/parquetIO.py (hunk 4 above)
Comment (ronawho) [reply benchmarks/parquetIO.py:6]: Based on https://github.com/bmcdonald3/arkouda/pull/2/files#diff-394b8a344150500f3214611a8f8716ecb550bf84cb725c8ff705fec3437ff952R1304, I think only `int64` today

File: src/ArrowInclude.chpl [hunk 9]
```diff
@@ -0,0 +1,101 @@
+module ArrowInclude {
+  use SysCTypes, CPtr, Time;
+  require "ArrowFunctions.h";
+  require "ArrowFunctions.o";
+
+  const ROWGROUPS = 512*1024*128; // 512 mb of int64
```
Comment (ronawho) [inline src/ArrowInclude.chpl:6]: Might be worth making this a `private config const` so we can override without having to rebuild. And I might write this as `512*1024*1024 / numBytes(int)`

File: Makefile [hunk 10]
```diff
@@ -22,15 +22,16 @@ else
 CHPL_FLAGS += --fast
 endif
 CHPL_FLAGS += -smemTrack=true
-CHPL_FLAGS += -lhdf5 -lhdf5_hl -lzmq
+CHPL_FLAGS += -lhdf5 -lhdf5_hl -lzmq -lparquet -larrow
```
Comment (ronawho) [inline Makefile:25]: TODO for @ronawho -- conditionalize this and other parquet flags based on `ARKOUDA_SERVER_PARQUET_SUPPORT` or something.

File: Makefile [hunk 11]
```diff
@@ -22,15 +22,16 @@ else
...
 
 # We have seen segfaults with cache remote at some node counts
 CHPL_FLAGS += --no-cache-remote
 
 # add-path: Append custom paths for non-system software.
 # Note: Darwin `ld` only supports `-rpath <path>`, not `-rpath=<paths>`.
 define add-path
-CHPL_FLAGS += -I$(1)/include -L$(1)/lib --ldflags="-Wl,-rpath,$(1)/lib"
+CHPL_INCLUDES += -I$(1)/include -L$(1)/lib$(2) 
+CHPL_FLAGS += $(CHPL_INCLUDES) --ldflags="-Wl,-rpath,$(1)/lib$(2)"
```
Comment (ronawho) [inline Makefile:34]: TODO for @ronawho  -- figure out how to do this without breaking existing `add-path` support. Maybe `add-path-64` or something (or see if we can just have `$(2)` be optional and if it's not set default to the current behavior)

File: Makefile [hunk 12]
```diff
@@ -85,8 +86,23 @@ install-hdf5:
...
-
+	echo '$$(eval $$(call add-path,$(HDF5_INSTALL_DIR),))' >> Makefile.paths
+
+ARROW_VER := 5.0.0
+ARROW_NAME_VER := apache-arrow-$(ARROW_VER)
+ARROW_FULL_NAME_VER := arrow-apache-arrow-$(ARROW_VER)
+ARROW_BUILD_DIR := $(DEP_BUILD_DIR)/$(ARROW_FULL_NAME_VER)
+ARROW_INSTALL_DIR := $(DEP_INSTALL_DIR)/arrow-install
+ARROW_LINK := https://github.com/apache/arrow/archive/refs/tags/$(ARROW_NAME_VER).tar.gz
+ARROW_OPTIONS := -DARROW_JEMALLOC=OFF -DARROW_COMPUTE=OFF -DARROW_IPC=OFF -DARROW_WITH_RE2=OFF -DARROW_WITH_SNAPPY=ON -DCMAKE_CXX_COMPILER=g++ -DCMAKE_C_COMPILER=gcc
```
Comment (ronawho) [inline Makefile:97]: TODO for @ronawho -- revisit which options we disable and how CC/CXX are set.

File: Makefile [hunk 13]
```diff
@@ -140,6 +161,14 @@ check-hdf5: $(HDF5_CHECK)
 	$(DEP_INSTALL_DIR)/$@ -nl 1
 	@rm -f $(DEP_INSTALL_DIR)/$@ $(DEP_INSTALL_DIR)/$@_real
 
+ARROW_CHECK = $(DEP_INSTALL_DIR)/checkArrow.chpl
+check-arrow: $(ARROW_CHECK)
+	@echo "Checking for Arrow"
+	g++ -O3 -std=c++11 -c $(ARKOUDA_SOURCE_DIR)/$(ARROW_CPP).cpp -o $(ARKOUDA_SOURCE_DIR)/$(ARROW_CPP).o $(CHPL_INCLUDES) -lparquet -larrow
```
Comment (ronawho) [inline Makefile:167]: Instead of copying this code, we should make this step depend on `compile-arrow-cpp`

File: src/ArrowFunctions.cpp [hunk 14]
```diff
@@ -0,0 +1,147 @@
+#include "ArrowFunctions.h"
+
+#include <iostream>
+#include <arrow/api.h>
+#include <arrow/io/api.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/arrow/writer.h>
+#include <parquet/exception.h>
+#include <parquet/api/reader.h>
```
Comment (ronawho) [inline src/ArrowFunctions.cpp:9]: Do you think we need all these includes? (Doesn't really matter, just curious if we're brining in more than we actually need)

File: src/ArrowFunctions.cpp [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+using namespace std;
```
Comment (ronawho) [inline src/ArrowFunctions.cpp:11]: Why is this needed (not very familiar with C++ personally, so may be naive question)

File: src/ArrowFunctions.cpp [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+
+int cpp_getSize(const char* filename) {
+  std::shared_ptr<arrow::io::ReadableFile> infile;
+  PARQUET_ASSIGN_OR_THROW(
```
Comment (ronawho) [inline src/ArrowFunctions.cpp:15]: What does `PARQUET_ASSIGN_OR_THROW` do? If it throws are we catching that somehow and returning it as an error message?

File: src/ArrowFunctions.cpp [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+#include <arrow/api.h>
+#include <arrow/io/api.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/arrow/writer.h>
+#include <parquet/exception.h>
+#include <parquet/api/reader.h>
+
+using namespace std;
+
+int cpp_getSize(const char* filename) {
```
Comment (ronawho) [inline src/ArrowFunctions.cpp:13]: Do you like the name `getSize` or do you think we should call this `getNumRows` or something?

File: src/ArrowInclude.chpl [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  proc getArrSize(filename: string) {
+    var size = c_getSize(filename.c_str());
+    return size;
+  }
+
+  proc getArrType(filename: string, colname: string) {
+    // TODO: throw error if type not 0 or 1
+    var arrType = c_getType(filename.c_str(), colname.c_str());
```
Comment (ronawho) [inline src/ArrowInclude.chpl:73]: When passing chapel strings to extern routines you need to do `string.localize().c_str()` to localize the string if it's remote.

File: src/ArrowInclude.chpl [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+        var locDom = A.localSubdomain();
+        var locArr = A[locDom];
+        c_writeColumnToParquet(myFilename.c_str(), c_ptrTo(locArr), 0, dsetname.c_str(), locDom.size, rowGroupSize);
+      }
+  }
+
+  proc write1DDistArrayParquet(filename: string, dsetname, A) {
+    var prefix = filename;
+    var extension = ".parquet";
```
Comment (ronawho) [inline src/ArrowInclude.chpl:97]: Looks like prefix and extension are unused?

File: src/ArrowInclude.chpl [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+    return {low..high by stride};
+  }
+  
+  proc readFilesByName(A, filenames: [] string, sizes: [] int, dsetname: string) {
+    var (subdoms, length) = getSubdomains(sizes);
+
+    coforall loc in A.targetLocales() do on loc {
+      var locFiles = filenames;
+      var locFiledoms = subdoms;
+      forall (filedom, filename) in zip(locFiledoms, locFiles) {
```
Comment (ronawho) [inline src/ArrowInclude.chpl:53]: I know we talked about eventually making operations parallel for speed improvements, but I think to start I would keep this serial.

File: setup.py (hunk 1 above)
Comment (bmcdonald3) [reply setup.py:140]: Changed

File: benchmarks/parquetIO.py (hunk 4 above)
Comment (bmcdonald3) [reply benchmarks/parquetIO.py:6]: `int64` for both and `int32` for reading, but not writing yet, but I don't think it's made it into the server code all the way... will fix that

This was something I wasn't sure how to handle because we could add support for pretty much all the int-related Arrow types and it shouldn't be difficult, but I wasn't sure which ones we wanted to include. Asking in the Arkouda slack channel.

File: arkouda/pdarrayclass.py (hunk 5 above)
Comment (bmcdonald3) [reply arkouda/pdarrayclass.py:1024]: No, thanks for pointing that out, I actually thought that I did but obviously don't after looking at the code. There is a `processFilenames()` function for HDF5 that I guess I forgot to add for the Parquet equivalent that I will add in and that actually should handle a lot of the error checking.

File: arkouda/pdarrayIO.py (hunk 6 above)
Comment (bmcdonald3) [reply arkouda/pdarrayIO.py:136]: Ya, but not much since I'm throwing like 2 errors total now for Parquet, but the logic is in there for it's use

File: arkouda/pdarrayIO.py (hunk 8 above)
Comment (bmcdonald3) [reply arkouda/pdarrayIO.py:185]: We support 1 and 2 and there are tests for both of those in `tests/parquet_test.py`. I figured I would leave in the comment and logic for the strings case since we eventually hope to extend the code to support that as well, but do you think we should remove that for now?

File: src/ArrowFunctions.cpp (hunk 14 above)
Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:9]: I thought that we did based on the Arrow docs, but doing some testing, we didn't need the last 2.

File: src/ArrowFunctions.cpp (hunk 15 above)
Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:11]: Do you just mean the `using namespace std;`? I was adding `cout`s for testing and this makes it less annoying to write those, but removed it now

File: src/ArrowFunctions.cpp (hunk 16 above)
Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:15]: I haven't found any helpful docs on this, but this is the most helpful piece of info I've found: https://arrow.apache.org/cookbook/cpp/basic.html, but it's on a different macro which I assume behaves the same.

It says that an error is returned if it fails, but I don't really know how to test that and I have never worked with errors in C++ and wouldn't know what to do with it without looking into it more. Will plan to explore this further as part of this sprints improve Arrow error handling tasks.

Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:15]: FWIW, there are never any additional error handling things in the example C++ code that uses these macros, they just do that and then move on without a try/catch block or anything like that

File: src/ArrowFunctions.cpp (hunk 17 above)
Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:13]: I'll change it to `getNumRows`

File: dep/checkArrow.chpl (hunk 2 above)
Comment (bmcdonald3) [reply dep/checkArrow.chpl:3]: I just have always done this and never questioned it, but testing it now makes me think that we need both of them, removing either of them causes errors. Were you thinking we would only need one? If so, which one?

File: benchmarks/parquetIO.py (hunk 4 above)
Comment (bmcdonald3) [reply benchmarks/parquetIO.py:6]: Thinking about this a little more, we will only be able to support int64 until we get the type checking working in the Arkouda side of things. I have a potential fix in mind that I am going to try now, and I have more confidence in this new idea than the previous ones (which there have been many)

**Comment by bmcdonald3** [comment]
Oh boy, rookie mistake -- that bug (that I have now spent like 2 days working on) was due to me passing in the wrong default dataset name... I guess this is a good example of why I should have started with the error handling

File: src/ArrowInclude.chpl [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+            c_readColumnByName(filename.localize().c_str(), c_ptrTo(col), dsetname.localize().c_str(), filedom.size);
+            A[filedom] = col;
+          }
+        }
+      }
+    }
+  }
+
+  proc getArrSize(filename: string) {
+    var size = c_getNumRows(filename.c_str());
```
Comment (ronawho) [inline src/ArrowInclude.chpl:67]: Missing `.localize()`

File: src/ArrowInclude.chpl [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+      var suffix = i: string;
+      filenames[i] = filename + "_LOCALE" + suffix + ".parquet";
+    }
+
+    coforall (loc, idx) in zip(A.targetLocales(), filenames.domain) do on loc {
+        const myFilename = filenames[idx];
+
+        var locDom = A.localSubdomain();
+        var locArr = A[locDom];
+        c_writeColumnToParquet(myFilename.c_str(), c_ptrTo(locArr), 0, dsetname.localize().c_str(), locDom.size, rowGroupSize);
```
Comment (ronawho) [inline src/ArrowInclude.chpl:92]: Missing `.localize()`

Comment (ronawho) [reply src/ArrowInclude.chpl:92]: I guess here we actually know it's local since it's created above, but I'd probably do it still just for consistency

File: src/ArrowFunctions.cpp (hunk 14 above)
Comment (ronawho) [reply src/ArrowFunctions.cpp]: exception.h seems to provide `PARQUET_ASSIGN_OR_THROW` so I'm surprised that's not needed

File: dep/checkArrow.chpl (hunk 2 above)
Comment (ronawho) [reply dep/checkArrow.chpl:3]: I don't see why both would be needed, this should just be a normal chapel file. Maybe we need to get a `-M` or something on the command line, but I can't see why we'd need to `require` a chpl file.

File: benchmarks/parquetIO.py (hunk 4 above)
Comment (ronawho) [reply benchmarks/parquetIO.py:6]: Arkouda pdarray doesn't support `int32` so here I'm really talking about the arkouda types, not the types stored in the file.

File: arkouda/pdarrayIO.py (hunk 8 above)
Comment (ronawho) [reply arkouda/pdarrayIO.py:185]: They should be marked with TODOs I think

File: src/ArrowFunctions.cpp (hunk 16 above)
Comment (ronawho) [reply src/ArrowFunctions.cpp:11]: We should see what happens when this errors. Maybe try passing in a filename that doesn't exist or something and see what happens?

File: benchmarks/parquetIO.py [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+    if args.correctness_only:
+        for dtype in TYPES:
+            check_correctness(dtype, args.path, args.seed)
+        sys.exit(0)
+    
+    print("array size = {:,}".format(args.size))
+    print("number of trials = ", args.trials)
+    time_ak_write_read(args.size, args.trials, args.dtype, args.path, args.seed)
+    sys.exit(0)
```
Comment (ronawho) [inline benchmarks/parquetIO.py:77]: Can you add this benchmark to the list of benchmarks to run in `benchmarks/run_benchmarks.py`?

**Comment by ronawho** [comment]
I think the PR message should be updated to make it more clear that this is initial support for ints only. I'd probably start the description with that saying:

> This adds the ability to read and write int64 Parquet files in Arkouda through `pdarray.save_parquet()`...

I know you said this at the end too, but I think it's worth making it clear right off the bat

---

I also think we should get CI testing working. This will be a good way to keep track of whether new commits are breaking things. For the `Install dependencies` sections in `.github/workflows/CI.yml` you'll either need to figure out the right `apt-get install` packages, or maybe just install `cmake` and do a `make install-parquet` step manually to start if the apt packages aren't obvious.


---

It also looks like you might have merge conflicts to resolve?

File: benchmarks/parquetIO.py (hunk 4 above)
Comment (bmcdonald3) [reply benchmarks/parquetIO.py:6]: Oh, I see, ya it's only int64 then

File: dep/checkArrow.chpl (hunk 2 above)
Comment (ronawho) [reply dep/checkArrow.chpl:3]: That's what `-M` is for

Comment (bmcdonald3) [reply dep/checkArrow.chpl:3]: You would prefer a `-M` to a require?

Comment (ronawho) [reply dep/checkArrow.chpl:3]: Yeah, using a require statement for this feels like abusing the feature to me.

Comment (bmcdonald3) [reply dep/checkArrow.chpl:3]: Do you think the `require "ArrowFunctions.h;` and `require "ArrowFunctions.cpp;` are alright in the `ArrowInclude.chpl` file?

Is this what you would consider a proper use of the feature or what would that be?

Comment (ronawho) [reply dep/checkArrow.chpl:3]: Require for C dependencies is fine and I think that's what the feature was originally added for. I think the abuse is using it for chpl files

Comment (ronawho) [reply dep/checkArrow.chpl:3]: I'm also not overly familiar with require statements, so might be worth double checking my thinking with Brad (original implementer) or somebody else.

File: src/ArrowFunctions.cpp (hunk 16 above)
Comment (bmcdonald3) [reply src/ArrowFunctions.cpp:11]: It gives one of these weird errors, which I had seen before and not thought about much:

```
libc++abi.dylib: terminating with uncaught exception of type parquet::ParquetStatusException: IOError: Failed to open local file 'asd.asd'. Detail: [errno 2] No such file or directory
```

We definitely want to be catching/handling that error, my initial thought is to just catch that error in the C++ code and return a status code and then raise the error on the Chapel side. Is that what you had in mind as well?

Comment (ronawho) [reply src/ArrowFunctions.cpp:11]: yeah, I think we need to catch and return as an error code.

Comment (ronawho) [reply src/ArrowFunctions.cpp:11]: You should look at the other macros in `exceptions.h`. There may already be other ways of returning an error code / status instead of manually catching errors. I think `PARQUET_ASSIGN_OR_THROW` is taking a status code and turning it into an exception, so there may be one that just gives you the status code (or maybe you just don't need the macro at all and you just return the return value of the thing you're calling)

Comment (ronawho) [reply src/ArrowFunctions.cpp:11]: See also https://arrow.apache.org/docs/cpp/api/support.html#error-return-and-reporting

File: .github/workflows/CI.yml [hunk 24]
```diff
@@ -122,7 +127,13 @@ jobs:
...
         apt-get update && apt-get install -y libhdf5-dev hdf5-tools libzmq3-dev python3-pip
+        apt-get install -y -V ca-certificates lsb-release wget
+        wget https://apache.jfrog.io/artifactory/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
+        apt-get install -y -V ./apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
+        apt-get update
+        apt-get install -y -V libarrow-dev libparquet-dev
         echo "\$(eval \$(call add-path,/usr/lib/x86_64-linux-gnu/hdf5/serial/))" > Makefile.paths
     - name: Run unit tests
       run: |
+        make compile-arrow-cpp
```
Comment (ronawho) [inline .github/workflows/CI.yml:138]: Let's move this logic into a skipif (if the `make` fails, then skip)

File: .github/workflows/CI.yml [hunk 25]
```diff
@@ -122,7 +127,13 @@ jobs:
     - name: Install dependencies
       run: |
         apt-get update && apt-get install -y libhdf5-dev hdf5-tools libzmq3-dev python3-pip
+        apt-get install -y -V ca-certificates lsb-release wget
+        wget https://apache.jfrog.io/artifactory/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
+        apt-get install -y -V ./apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
+        apt-get update
+        apt-get install -y -V libarrow-dev libparquet-dev
```
Comment (ronawho) [inline .github/workflows/CI.yml:134]: @ronawho simplify install step

File: Makefile [hunk 26]
```diff
@@ -22,15 +22,16 @@ else
...
 
 # We have seen segfaults with cache remote at some node counts
 CHPL_FLAGS += --no-cache-remote
 
 # add-path: Append custom paths for non-system software.
 # Note: Darwin `ld` only supports `-rpath <path>`, not `-rpath=<paths>`.
 define add-path
-CHPL_FLAGS += -I$(1)/include -L$(1)/lib --ldflags="-Wl,-rpath,$(1)/lib"
+CHPL_INCLUDES += -I$(1)/include -L$(1)/lib$(2) 
+CHPL_FLAGS += -I$(1)/include -L$(1)/lib$(2) --ldflags="-Wl,-rpath,$(1)/lib$(2)"
```
Comment (ronawho) [inline Makefile:34]: @ronawho -- simplify this logic and avoid breaking existing `add-path` uses.

File: arkouda/pdarrayclass.py (hunk 5 above)
Comment (ronawho) [reply arkouda/pdarrayclass.py:1061]: @bmcdonald3 -- change this to only support truncate and throw error if append is used (and add TODO to support this mode)

File: benchmarks/run_benchmarks.py [hunk 27]
```diff
@@ -22,7 +22,7 @@
 BENCHMARKS = ['stream', 'argsort', 'coargsort', 'groupby', 'aggregate', 'gather', 'scatter',
               'reduce', 'scan', 'noop', 'setops', 'array_create',
               'array_transfer', 'IO', 'str-argsort', 'str-coargsort',
-              'str-groupby', 'str-gather', 'substring_search', 'flatten']
+              'str-groupby', 'str-gather', 'substring_search', 'flatten', 'parquetIO']
```
Comment (ronawho) [inline benchmarks/run_benchmarks.py:25]: For now we'll only add `parquetIO` if env var is set, later on though we should do something to query if the server was built with parquet support and auto-skip the benchmark.

File: tests/parquet_test.py [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+    for i in range(NUMFILES):
+        sz = len(test_arrs[i])
+        failures += compare_values(test_arrs[i], pq_arr[(i*sz):(i*sz)+sz])
+    for f in glob.glob('pq_test*'):
+        os.remove(f)
+    return failures
+
+
+class ParquetTest(ArkoudaTest):
```
Comment (ronawho) [inline tests/parquet_test.py:46]: We'll want to skip this test if parquet env var isn't set (and longer term doing get_config auto-detection). https://stackoverflow.com/questions/22603562/how-to-check-a-member-variable-with-py-test-skipif

File: test/UnitTestParquetCpp.chpl [hunk 29]
```diff
@@ -0,0 +1,73 @@
+require '../src/ArrowInclude.chpl';
```
Comment (ronawho) [inline test/UnitTestParquetCpp.chpl:1]: Remove require

File: src/ArrowFunctions.h [hunk 30]
```diff
@@ -0,0 +1,32 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+  #define ARROWINT64 0
+  #define ARROWINT32 1
+  #define ARROWUNDEFINED -1
```
Comment (ronawho) [inline src/ArrowFunctions.h:7]: undefined or currently unsupported?

File: src/ArrowInclude.chpl [hunk 31]
```diff
@@ -0,0 +1,106 @@
+module ArrowInclude {
```
Comment (ronawho) [inline src/ArrowInclude.chpl:1]: Rename to Parquet?

File: src/ArrowInclude.chpl [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+  use SysCTypes, CPtr, Time, IO, Reflection;
```
Comment (ronawho) [inline src/ArrowInclude.chpl:2]: Reflection and Time needed?

File: src/ArrowInclude.chpl [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+  require "ArrowFunctions.o";
+
+  private config const ROWGROUPS = 512*1024*1024 / numBytes(int); // 512 mb of int64
+  
+  extern proc c_getNumRows(chpl_str): int;
+  extern proc c_readColumnByName(filename, chpl_arr, colNum, numElems);
+  extern proc c_getType(filename, colname): c_int;
+  extern proc c_writeColumnToParquet(filename, chpl_arr, colnum,
+                                     dsetname, numelems, rowGroupSize);
+  extern proc c_getVersionInfo(): c_string;
```
Comment (ronawho) [inline src/ArrowInclude.chpl:14]: Let's move the extern proc's to the functions that actually use them (to avoid leaking these into the namespace)

File: src/ArrowInclude.chpl [hunk 34]
```diff
@@ -0,0 +1,106 @@
...
+  extern var ARROWINT64: c_int;
+  extern var ARROWINT32: c_int;
+  extern var ARROWUNDEFINED: c_int;
+  
+
+  enum ArrowTypes { int64, int32, notimplemented };
+  
+  proc getVersionInfo() {
+    extern proc strlen(str): c_int;
+    var cVersionString = c_getVersionInfo();
```
Comment (ronawho) [inline src/ArrowInclude.chpl:25]: Let's have `c_getVersionInfo` call `strdup`, and then do a `defer c_free(cVersionString:c_void_ptr)` or something here. (We think `strdup` might not be neede since this is probably global, but let's play it safe.)

File: src/GenSymIO.chpl [hunk 35]
```diff
@@ -1,4 +1,5 @@
 module GenSymIO {
+    use ArrowInclude as Arrow;
```
Comment (ronawho) [inline src/GenSymIO.chpl:2]: Remove this use

File: src/GenSymIO.chpl [hunk 36]
```diff
@@ -644,6 +645,159 @@ module GenSymIO {
         return new MsgTuple(repMsg,MsgType.NORMAL);
     }
 
+    proc readAllParquetMsg(cmd: string, payload: string, st: borrowed SymTab): MsgTuple throws {
```
Comment (ronawho) [inline src/GenSymIO.chpl:648]: Add a config param (Maybe `hasParquetSupport`) and conditionalize this function based on that param (returning an error if it's not supported)

File: src/GenSymIO.chpl [hunk 37]
```diff
@@ -1235,6 +1389,61 @@ module GenSymIO {
         }
     }
 
+    proc toparquetMsg(cmd: string, payload: string, st: borrowed SymTab): MsgTuple throws {
```
Comment (ronawho) [inline src/GenSymIO.chpl:1392]: Add param conditional and use stmt here (like for read function)

File: .github/workflows/CI.yml (hunk 24 above)
Comment (bmcdonald3) [reply .github/workflows/CI.yml:138]: Since $CHPL_FLAGS has `-larrow -lparquet`, we need this make compile-arrow-cpp for every test with how things are today, so we could either change the `Makefile` to not add those flags onto everything or just leave this `make compile-arrow-cpp` here for now and come back to that in the future. Since what you were saying about this likely not seeing any use by anyone except CI, I vote to just keep it like this.

File: Makefile [hunk 38]
```diff
@@ -374,6 +406,7 @@ $(TEST_BINARY_DIR):
 
 .PHONY: $(TEST_TARGETS) # Force tests to always rebuild.
 $(TEST_TARGETS): $(TEST_BINARY_DIR)/$(TEST_BINARY_SIGIL)%: $(TEST_SOURCE_DIR)/%.chpl | $(TEST_BINARY_DIR)
+	make compile-arrow-cpp
```
Comment (ronawho) [inline Makefile:409]: Let's remove this and the one above for test-chapel

Comment (bmcdonald3) [reply Makefile:411]: So, that means that in order for either of those to work, the user would have to run a regular `make`, since the C++ code isn't being built elsewhere, right? (Not that I have a problem with that, just making sure I understand)

Comment (ronawho) [reply Makefile]: They'd have to run `make compile-arrow-cpp`, yeah. This isn't an ideal solution, but I'm not sure what a better way is at the moment. Cause if they don't have arrow support, the test test will fail, so maybe we'd need to do something to filter the arrow test out of this list or something?

Comment (bmcdonald3) [reply Makefile]: One possible solution could be to create a `test/arrow/` dir and put the arrow tests in there and we could add a `ARROW_TESTS` flag that would add it onto the start_test run
